### DISCUSSION TITLE
When does always transition occur?

### DISCUSSION THREAD

**Discussion by tinitus-git** [statelyai/xstate#3034] [Q&A]
I have created the following simple machine:

`import { createMachine } from "xstate";

const genericMachine = createMachine(
  {
    id: "FsmTypeMachine",
    context: {
      StateVar: "",
    },
    initial: "State1",
    states: {
      State1: {
        always: {
          target: "State2",
          cond: "StateVarCondition",
        },
        on: {
          AssignValue: {
            actions: "AssignStateVar",
          },
        },
      },
      State2: {
        type: "final",
      },
    },
  },
  {
    actions: {
      AssignStateVar: (context, event) => {
        context.StateVar = event.data;
      }
    },
    guards: {
      StateVarCondition: (context) => {
        return context.StateVar !== "";
      },
    },
  }
);`

As you can see, the machine consists of two states. State1 has an always transition (based on a context variable) and an event to assign the context variable. When I send the following event through the visualizer:

{
"type":"AssignValue"
"data":"SomeValue"
}

I see that the event gets sent, the context variable is assigned, but the machine does not transition to State2 via the always transition. I was hoping that the AssignValue event would assign the value through the transition defined in the "on" which re-enters the State1 state and the always transition would be triggered, but that does not seem to be the case.

I have come up with a number of work arounds, but I do not know if this is the way things are supposed to work or not.

Thanks,

Brian

**Comment by jmakeig** [comment]
You need to use the [`assign()`](https://xstate.js.org/docs/guides/actions.html#api) action creator for your context update:

```js
AssignStateVar: assign({StateVar: (context, event) => event.data})
```
